code/modules/antagonists/_common/antag_hud.dm

```diff
@@ -1,58 +1,78 @@
-/datum/mind/proc/transfer_antag_huds(datum/atom_hud/antag/newhud)
-	leave_all_antag_huds()
-	set_antag_hud(current, antag_hud_icon_state)
-	if(newhud)
-		newhud.join_hud(current)
-
-/datum/mind/proc/leave_all_antag_huds()
-	for(var/datum/atom_hud/antag/hud in GLOB.huds)
-		if(hud.hudusers[current])
-			hud.leave_hud(current)
+/// An alternate appearance that will only show if you have the antag datum
+/datum/atom_hud/alternate_appearance/basic/has_antagonist
+	var/antag_datum_type
+
+/datum/atom_hud/alternate_appearance/basic/has_antagonist/New(key, image/I, antag_datum_type)
+	src.antag_datum_type = antag_datum_type
+	return ..(key, I, NONE)
+
+/datum/atom_hud/alternate_appearance/basic/has_antagonist/mobShouldSee(mob/M)
+	return !!M.mind?.has_antag_datum(antag_datum_type)
+
+/// An alternate appearance that will show all the antagonists this mob has
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud
+	var/list/antag_hud_images = list()
+	var/index = 1
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/New(key, datum/mind/mind)
+	antag_hud_images = get_antag_hud_images(mind)
+
+	var/image/first_antagonist = get_antag_image(1) || image(icon('icons/blanks/32x32.dmi', "nothing"), mind.current)
+
+	RegisterSignal(
+		mind,
+		list(COMSIG_ANTAGONIST_GAINED, COMSIG_ANTAGONIST_REMOVED),
+		.proc/update_antag_hud_images
+	)
+
+	check_processing()
+
+	return ..(key, first_antagonist, NONE)
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/Destroy()
+	QDEL_LIST(antag_hud_images)
+	STOP_PROCESSING(SSantag_hud, src)
+
+	return ..()
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/mobShouldSee(mob/mob)
+	return Master.current_runlevel >= RUNLEVEL_POSTGAME || mob.client?.combo_hud_enabled
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/process(delta_time)
+	index += 1
+	update_icon()
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/proc/check_processing()
+	if (antag_hud_images.len > 1 && !(DF_ISPROCESSING in datum_flags))
+		START_PROCESSING(SSantag_hud, src)
+	else if (antag_hud_images.len <= 1)
+		STOP_PROCESSING(SSantag_hud, src)
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/proc/get_antag_image(index)
+	RETURN_TYPE(/image)
+	if (antag_hud_images.len)
+		return antag_hud_images[(index % antag_hud_images.len) + 1]
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/proc/get_antag_hud_images(datum/mind/mind)
+	var/list/final_antag_hud_images = list()
+
+	for (var/datum/antagonist/antagonist as anything in mind?.antag_datums)
+		if (isnull(antagonist.antag_hud_name))
+			continue
+		final_antag_hud_images += image('icons/mob/hud.dmi', mind.current, antagonist.antag_hud_name)
+
+	return final_antag_hud_images
+
+/datum/atom_hud/alternate_appearance/basic/antagonist_hud/proc/update_icon()
+	if (antag_hud_images.len == 0)
+		image.icon = icon('icons/blanks/32x32.dmi', "nothing")
+	else
+		image.icon = icon('icons/mob/hud.dmi', get_antag_image(index).icon_state)
```

Comment:
Just as a small suggestion, can the dmi file be de-hardcoded?

Reply:
i kinda like the idea of making antag huds their own dmi instead